File: 백준/Silver/1074. Z/Z.java
import java.util.Scanner;

public class Main {
	static long cnt = 0;
	static int R;
	static int C;
	static boolean flag;
	public static void main(String[] args) {
		Scanner sc = new Scanner(System.in);
		
		int N = sc.nextInt();
		
		R = sc.nextInt();
		 
		C = sc.nextInt();
				
		int L = (int) Math.pow(2, N);
		
		find(R, C, L, 0);
		
		
	}
	static void find(int r , int c, int len, int num) {
		
		if(len == 1) {

			if(r==1) {
				System.out.println((int)(num+r+c+1));
				return;
			}
			System.out.println((int)(num+r+c));
			return;
		}
		
		
		if(r>=len/2) {
			num += ((len*len)/4) *2;
			r-=len/2;
		}
		if(c>=len/2) {
			num += (len*len)/4;
			c-=len/2;
		}
		find(r, c, len/2, num);
		
	}
	


}

File: 백준/Silver/1074. Z/Z.py
def split_recursion(l, i, j):
    global cnt,flag
    if flag:
        return
    if l == 1:
        if i==r and j==c:
            flag = True
            return
        cnt+=1

        return
    size = l//2
    #제1사분면
    if i<=r<i+size and j<=c<j+size:
        split_recursion(size, i, j)
    elif i<=r<i+size and j+size<=c:
        cnt+=size*size
        split_recursion(size, i, j+size)
    elif i+size<=r and j<=c<j+size:
        cnt+=size*size*2
        split_recursion(size, i+size, j)
    else:
        cnt+=size*size*3
        split_recursion(size, i+size, j+size)

n, r, c = map(int, input().split())
cnt = 0
flag = False
split_recursion(2**n, 0, 0)
print(cnt)